### PR TITLE
add a new input format vidarr_provenance to vidarr plugin

### PR DESCRIPTION
**Pull request by grapigeau**
JIRA Ticket: https://jira.oicr.on.ca/browse/GP-4528

- [x] Updates Changelog
- [ ] Updates developer documentation

### CODE REVIEW

File: docs/plugin-vidarr.md [hunk 1]
```diff
@@ -18,3 +18,22 @@ should be submitted to Vidarr. If null, worklow runs will always be submitted.
 If a number, in seconds, is given, then the action must have been
 (re-)generated by an olive in that window to be submitted to Vidarr. If that
 window is exceeded, they will be run as dry-run.
+
+The vidarr plugin provides an imput format: `vidarr_provenance` which
```
Comment (avarsava) [inline docs/plugin-vidarr.md:22]: imput -> input

File: docs/plugin-vidarr.md [hunk 2]
```diff
@@ -18,3 +18,22 @@ should be submitted to Vidarr. If null, worklow runs will always be submitted.
 If a number, in seconds, is given, then the action must have been
 (re-)generated by an olive in that window to be submitted to Vidarr. If that
 window is exceeded, they will be run as dry-run.
+
+The vidarr plugin provides an imput format: `vidarr_provenance` which 
+provides the analyis provenance data.
+To use, add to the configuration file:
```
Comment (avarsava) [inline docs/plugin-vidarr.md:24]: 'add to the configuration file' suggests adding this block as secondary to the block defined at line 8. Could these two sections be combined?

File: docs/plugin-vidarr.md [hunk 3]
```diff
@@ -18,3 +18,22 @@ should be submitted to Vidarr. If null, worklow runs will always be submitted.
...
+      "defaultMaxSubmissionDelay": null,
+      "url": "http://vidarr:8000",
+      "instance": "development",
+      "analysisTypes": ["FILE"],
+      "versionTypes": []
+    }
+
+The `"instance"` is used to determine workflow accession values.
+
+The `"analysisTypes"` specifies the type of analysis desired.
```
Comment (avarsava) [inline docs/plugin-vidarr.md:36]: What are the available options and how do they affect the content of the input format?

File: docs/plugin-vidarr.md [hunk 4]
```diff
@@ -18,3 +18,22 @@ should be submitted to Vidarr. If null, worklow runs will always be submitted.
...
+
+    {
+      "defaultMaxSubmissionDelay": null,
+      "url": "http://vidarr:8000",
+      "instance": "development",
+      "analysisTypes": ["FILE"],
+      "versionTypes": []
+    }
+
+The `"instance"` is used to determine workflow accession values.
```
Comment (avarsava) [inline docs/plugin-vidarr.md:34]: I'm not sure what this means. Each URL:port combo can only have one instance.

File: docs/plugin-vidarr.md (hunk 1 above)
Comment (avarsava) [inline docs/plugin-vidarr.md:22]: vidarr -> Vidarr

File: docs/plugin-vidarr.md [hunk 5]
```diff
@@ -18,3 +18,22 @@ should be submitted to Vidarr. If null, worklow runs will always be submitted.
...
+
+The `"analysisTypes"` specifies the type of analysis desired. 
+
+The `"versionTypes"` specifies the version keys that should be returned; if empty,
```
Comment (avarsava) [inline docs/plugin-vidarr.md:38]: Why is it 'types' in the option and 'keys' in the description? Also what are these; is there a vidarr doc that could be linked?

File: plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrProvenanceDto.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+import ca.on.oicr.gsi.shesmu.plugin.Tuple;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/** Bean of provenance response 'results' from Vidarr */
```
Comment (avarsava) [inline plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrProvenanceDto.java:11]: This seems like a duplicate of https://github.com/oicr-gsi/vidarr/blob/master/vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/api/ProvenanceWorkflowRun.java ?

File: plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrProvenanceResponseDto.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
+package ca.on.oicr.gsi.shesmu.vidarr;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.JsonNode;
+import java.util.Set;
+
+/** Bean of provenance response from Vidarr */
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class VidarrProvenanceResponseDto {
```
Comment (avarsava) [inline plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrProvenanceResponseDto.java:10]: Seems like a duplicate of https://github.com/oicr-gsi/vidarr/blob/master/vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/api/AnalysisProvenanceResponse.java ?

File: docs/plugin-vidarr.md (hunk 4 above)
Comment (grapigeau) [reply docs/plugin-vidarr.md:26]: It's only relevant here: https://github.com/oicr-gsi/shesmu/blob/a69c6b72a083dbca92e203150355e64d88226ab1/plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java#L137. Is there a better way to access the instance name? I don't it's anywhere in the response.

File: docs/plugin-vidarr.md (hunk 5 above)
Comment (avarsava) [reply docs/plugin-vidarr.md:28]: Yeah I also can't find anything. I've looked at the code and I still don't understand what this means. This should be documented somewhere

File: plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java [hunk 8]
```diff
@@ -50,6 +54,99 @@
...
+      super("vidarr-provenance " + fileName.toString(), 30, ReplacingRecord::new);
+    }
+
+    private String formatListAsString(List<String> list) {
+      return list.stream().map(name -> ("\"" + name + "\"")).collect(Collectors.joining(","));
+    }
+
+    private String createVidarrProvenanceRequestBody(List<String> versionTypes) {
+      return "{"
+          + "\"analysisTypes\": [\"FILE\"],"
```
Comment (avarsava) [inline plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java:68]: Why the constraint of only reporting FILE? We don't really use URL right now but we might in the future.
Also, this constraint should be documented.

File: docs/plugin-vidarr.md [hunk 9]
```diff
@@ -7,7 +7,8 @@ To integrate Shesmu with a Vidarr server, create a configuration file ending in
 
     {
       "defaultMaxSubmissionDelay": null,
-      "url": "http://vidarr:8000"
+      "url": "http://vidarr:8000",
+      "versionTypes": []
```
Comment (avarsava) [inline docs/plugin-vidarr.md:11]: I think the instance name is still needed in the config?

File: plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrProvenanceDefinition.java [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package ca.on.oicr.gsi.shesmu.vidarr;
+
+import ca.on.oicr.gsi.shesmu.plugin.input.InputFormat;
+
+public class VidarrProvenanceDefinition extends InputFormat {
+
+  public VidarrProvenanceDefinition() {
+    super("vidarr_analysis", VidarrProvenanceValue.class);
```
Comment (avarsava) [inline plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrProvenanceDefinition.java:8]: The input format is called 'vidarr_analysis' in shesmu by this line. Did you want to call it 'vidarr_provenance'?

Comment (callunity) [reply plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrProvenanceDefinition.java:8]: The other input format definition classes end with `XYFormatDefinition` . I do agree that it's unclear to have the class be Vidarr Provenance and the input format name be vidarr analysis; perhaps vidarr analysis for both? (so class becomes `VidarrAnalysisFormatDefinition`)

File: plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java [hunk 11]
```diff
@@ -50,6 +54,99 @@
...
+          + "\"timestamp\": 0,"
+          + "\"versionPolicy\": \"LATEST\","
+          + "\"versionTypes\": ["
+          + versionTypes.stream().map(name -> ("\"" + name + "\"")).collect(Collectors.joining(","))
+          + "  ]}";
+    }
+
+    protected Stream<VidarrAnalysisValue> provenanceArchive(String baseUrl) throws Exception {
+
+      if (configuration.isEmpty()) {
```
Comment (avarsava) [inline plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java:76]: the only place provenanceArchive() is called is in fetch(), which also tests for empty configuration. Don't need to test twice

File: plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java [hunk 12]
```diff
@@ -50,6 +54,99 @@
 
 public class VidarrPlugin extends JsonPluginFile<Configuration> {
 
+  private class ProvenanceCache extends ValueCache<Stream<VidarrAnalysisValue>> {
+    public ProvenanceCache(Path fileName) {
+      super("vidarr-provenance " + fileName.toString(), 30, ReplacingRecord::new);
```
Comment (callunity) [inline plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java:59]: What do you think about changing provenance -> analysis in these cases as well?

File: plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java [hunk 13]
```diff
@@ -50,6 +54,95 @@
...
+                                      analysisRecord.getLabels(),
+                                      analysisRecord.getSize(),
+                                      analysisRecord.getId(),
+                                      analysisRecord.getMetatype(),
+                                      analysisRecord.getPath()))
+                          .collect(Collectors.toSet()),
+                      ca.getExternalKeys().stream()
+                          .map(
+                              eKey ->
+                                  new Tuple(eKey.getId(), eKey.getProvider(), eKey.getVersions()))
```
Comment (avarsava) [inline plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java:125]: I tried `versionTypes:["nonsense"]` and simulations failed because of a deeply nested attempt to get the entryset of a null map. This may need to place an empty Map if eKey.getVersions is null

File: plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java (hunk 8 above)
Comment (callunity) [reply plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java:64]: I agree with Alexis about not constraining this, and it would be much easier to do so by updating a config file than having to do a Shesmu release.

File: plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java [hunk 14]
```diff
@@ -50,6 +54,95 @@
...
+          HTTP_CLIENT.send(
+              HttpRequest.newBuilder(URI.create(baseUrl + "/api/provenance"))
+                  .header("Content-type", "application/json")
+                  .timeout(Duration.ofMinutes(10))
+                  .POST(
+                      HttpRequest.BodyPublishers.ofString(
+                          createVidarrProvenanceRequestBody(configuration.get().getVersionTypes())))
+                  .build(),
+              new JsonBodyHandler<>(
+                  MAPPER, new TypeReference<AnalysisProvenanceResponse<ExternalKey>>() {}));
```
Comment (callunity) [inline plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java:90]: The [VersionPolicy documentation](https://github.com/oicr-gsi/vidarr/blob/master/vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/api/VersionPolicy.java#L14) says that the type argument for AnalysisProvenanceResponse should differ depending on the VersionPolicy. Unfortunately, the different types (ExternalId, ExternalKey, MultiVersionExternalKey) would need to have different ShesmuVariable types as they have different version representations, which gets messy really quickly. Given that we're expecting to always use the LATEST version policy, what do y'all think about hard-coding that (so, removing it from configuration) and keeping this type argument as ExternalKey?

Comment (avarsava) [reply plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java:90]: The version policy is already hardcoded: https://github.com/oicr-gsi/shesmu/pull/1350/files#diff-5e0eae5036eb27102f91d90877b5132685e42e8990c087c3bee9c73b7330dd73R68

File: docs/plugin-vidarr.md [hunk 15]
```diff
@@ -18,3 +19,8 @@ should be submitted to Vidarr. If null, worklow runs will always be submitted.
 If a number, in seconds, is given, then the action must have been
 (re-)generated by an olive in that window to be submitted to Vidarr. If that
 window is exceeded, they will be run as dry-run.
+
+The Vidarr plugin provides an input format: `vidarr_analysis` which provides the file type analyis provenance data.
```
Comment (callunity) [inline docs/plugin-vidarr.md:23]: Could you please update the docs now that analysisTypes is configurable?

File: plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java [hunk 16]
```diff
@@ -50,6 +54,107 @@
...
+      try {
+        final var body = results.body().get();
+        return body.getResults().stream()
+            .map(
+                ca ->
+                    new VidarrAnalysisValue(
+                        ca.getCompleted() == null
+                            ? Optional.empty()
+                            : Optional.of(ca.getCompleted().toInstant()),
+                        ca.getAnalysis().stream()
```
Comment (avarsava) [inline plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java:108]: If I make analysisTypes an empty array, this line attempts to call stream() on null. It doesn't really make sense to make analysisTypes empty, but there should be an earlier and clearer error in that case

File: docs/plugin-vidarr.md [hunk 17]
```diff
@@ -22,5 +23,7 @@ window is exceeded, they will be run as dry-run.
 
 The Vidarr plugin provides an input format: `vidarr_analysis` which provides the file type analyis provenance data.
 
+The `"analysisTypes"` specifies the type of analysis desired. The default is only files.
```
Comment (callunity) [inline docs/plugin-vidarr.md:26]: Probably worth linking to the [Vidarr doc for analysis type](https://github.com/oicr-gsi/vidarr/blob/757cdb311109aefa6bd650172bd78fed04507fb7/vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/api/AnalysisOutputType.java) since it's controlled vocabulary. 

Also, it doesn't look like "FILES" is set as a default anywhere in the code; only in your example, so please remove or rework the second sentence.

File: plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/Configuration.java [hunk 18]
```diff
@@ -1,8 +1,12 @@
 package ca.on.oicr.gsi.shesmu.vidarr;
 
+import java.util.List;
+
 public class Configuration {
   private Long defaultMaxSubmissionDelay;
   private String url;
+  private List<String> analysisTypes;
```
Comment (avarsava) [inline plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/Configuration.java:8]: What do you think about making this a List<AnalysisOutputType>  from pluginapi?

**Review by avarsava — DISMISSED** [review]

File: docs/plugin-vidarr.md [hunk 19]
```diff
@@ -23,7 +23,7 @@ window is exceeded, they will be run as dry-run.
 
 The Vidarr plugin provides an input format: `vidarr_analysis` which provides the file type analyis provenance data.
 
-The `"analysisTypes"` specifies the type of analysis desired. The default is only files.
+The `"analysisTypes"` specifies the type of analysis desired. For excepted values reference the [Vidarr docs](https://github.com/oicr-gsi/vidarr/blob/757cdb311109aefa6bd650172bd78fed04507fb7/vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/api/AnalysisOutputType.java).
```
Comment (callunity) [inline docs/plugin-vidarr.md:26]: excepted -> expected

File: docs/plugin-vidarr.md [hunk 20]
```diff
@@ -18,3 +20,14 @@ should be submitted to Vidarr. If null, worklow runs will always be submitted.
 If a number, in seconds, is given, then the action must have been
 (re-)generated by an olive in that window to be submitted to Vidarr. If that
 window is exceeded, they will be run as dry-run.
+
+The Vidarr plugin provides an input format: `vidarr_analysis` which provides the file type analyis
+provenance data.
+
+The `"analysisTypes"` specifies the type of analysis desired. For expected values reference
+the [Vidarr docs](https://github.com/oicr-gsi/vidarr/blob/757cdb311109aefa6bd650172bd78fed04507fb7/vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/api/AnalysisOutputType.java).
```
Comment (avarsava) [inline docs/plugin-vidarr.md:28]:
```suggestion
the [Vidarr AnalysisOutputType class](https://github.com/oicr-gsi/vidarr/blob/master/vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/api/AnalysisOutputType.java).
```

File: docs/plugin-vidarr.md [hunk 21]
```diff
@@ -18,3 +20,14 @@ should be submitted to Vidarr. If null, worklow runs will always be submitted.
 If a number, in seconds, is given, then the action must have been
 (re-)generated by an olive in that window to be submitted to Vidarr. If that
 window is exceeded, they will be run as dry-run.
+
+The Vidarr plugin provides an input format: `vidarr_analysis` which provides the file type analyis
```
Comment (avarsava) [inline docs/plugin-vidarr.md:24]: No longer only provides the file type analysis

File: plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java (hunk 16 above)
Comment (avarsava) [reply plugin-vidarr/src/main/java/ca/on/oicr/gsi/shesmu/vidarr/VidarrPlugin.java]: this is also true of setting analysisTypes to just 'URL' enough though that should be a valid configuration (just with no records)

File: docs/plugin-vidarr.md (hunk 21 above)
Comment (avarsava) [reply docs/plugin-vidarr.md:24]: also analyis -> analysis

**Comment by grapigeau** [comment]
Just FYI I find keeping the commit history easier for tracking during PR review. I will rebase before merging

File: docs/plugin-vidarr.md [hunk 22]
```diff
@@ -22,7 +20,14 @@ If a number, in seconds, is given, then the action must have been
...
 The Vidarr plugin provides an input format: `vidarr_analysis` which provides the analysis
-provenance data. 
+provenance data. To use this input format add the fields to the configuration file as follows:
+
+    {
+      "defaultMaxSubmissionDelay": null,
+      "url": "http://vidarr:8000",
+      "analysisTypes": ["FILE"],
+      "versionTypes": []
+    }
```
Comment (avarsava) [inline docs/plugin-vidarr.md:30]: Again, this could be mistaken for needing to be two parallel objects in the config file. 

```suggestion
provenance data. To use this input format add the fields to the configuration file as follows:

      "analysisTypes": ["FILE"],
      "versionTypes": []
```

**Review by callunity — APPROVED** [review]
